airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/state/MemoryManager.kt

```diff
@@ -19,17 +21,39 @@ import kotlinx.coroutines.sync.withLock
  * TODO: Some degree of logging/monitoring around how accurate we're actually being?
  */
 @Singleton
-class MemoryManager(private val availableMemoryProvider: AvailableMemoryProvider) {
-    private val totalMemoryBytes: Long = availableMemoryProvider.availableMemoryBytes
+class MemoryManager(private val totalMemoryBytes: Long) {
+    constructor(provider: AvailableMemoryProvider) : this(provider.availableMemoryBytes)
+
     private var usedMemoryBytes = AtomicLong(0L)
     private val mutex = Mutex()
     private val syncChannel = Channel<Unit>(Channel.UNLIMITED)
 
+    /**
+     * Releasable reservation of memory. For large blocks (ie, from [reserveRatio], provides a
+     * submanager that can be used to manage allocating the reservation).
+     */
+    data class Reservation(val memoryManager: MemoryManager, val bytes: Long) : CloseableCoroutine {
+        private var released = AtomicBoolean(false)
+
+        suspend fun release() {
+            if (!released.compareAndSet(false, true)) {
+                return
```

Comment:
I'd throw an exception here. It's pretty bad if this is happening

Reply:
The usual contract for `close` is that multiple calls conflate.

Reply:
yeah, that's right.